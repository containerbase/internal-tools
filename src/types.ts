export enum Commands {
  BinaryBuilder = 'binary-builder',
  DockerConfig = 'docker-config',
  DockerBuilder = 'docker-builder',
}
