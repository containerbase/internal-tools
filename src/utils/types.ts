export type ExecResult = {
  readonly code: number;
  readonly stdout: string;
  readonly stderr: string;
};

export class ExecError extends Error implements ExecResult {
  readonly name = 'ExecError';

  constructor(
    public readonly code: number,
    public readonly stdout: string,
    public readonly stderr: string,
    public readonly cmd: string
  ) {
    super(`ExecError: (${code}) ` + stderr.split('\n').slice(-10).join('\n'));
  }
}

export type ConfigFile = {
  datasource: string;
  image: string;
  depName?: string;
  versioning?: string;
  startVersion: string;
  cache?: string;
  buildArg?: string;
  ignoredVersions?: string[];
  forceUnstable?: boolean;
  versions?: string[];
  latestVersion?: string;
  maxVersions?: number;
  extractVersion?: string;
};

export type Config = {
  buildArg: string;
  buildArgs: string[];
  buildOnly: boolean;
  tagSuffix?: string;
  depName: string;
  imagePrefix: string;
  image: string;
  ignoredVersions: string[];
  majorMinor: boolean;
  lastOnly: boolean;
  dryRun: boolean;
  prune: boolean;
  platforms?: string[];
} & ConfigFile;

export type BinaryBuilderConfig = {
  buildArgs: string[];
  depName: string;
  ignoredVersions: string[];
  lastOnly: boolean;
  dryRun: boolean;
} & ConfigFile;
