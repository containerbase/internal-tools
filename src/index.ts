// Node v16 is missing structuredClone, so we polyfill it here.
import 'core-js/actual/structured-clone';
import run from './runner';

run().catch(console.error);
