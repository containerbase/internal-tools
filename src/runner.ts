import { getInput, setFailed } from '@actions/core';
import * as chalk from 'chalk';
import { Commands } from './types';
import log from './utils/logger';

export default async function run(): Promise<void> {
  try {
    log.info(chalk.blue('Renovate Docker Builder'));
    const cmd = getInput('command') as Commands;
    log.info(chalk.yellow('Executing:'), ` ${cmd}`);
    switch (cmd) {
      case Commands.BinaryBuilder:
        await (await import('./commands/binary')).run();
        break;
      case Commands.DockerBuilder:
        await (await import('./commands/docker/builder')).run();
        break;

      case Commands.DockerConfig:
        await (await import('./commands/docker/config')).run();
        break;

      default:
        log.error(chalk.red('Unknown command:'), cmd);
        break;
    }
  } catch (error) {
    setFailed((error as Error).message);
  }
}
