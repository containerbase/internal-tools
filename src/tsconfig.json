{
  "extends": "../tsconfig.json",
  "compilerOptions": {
    "rootDir": ".",
    "outDir": "dist",
    "types": ["node"]
  }
}
