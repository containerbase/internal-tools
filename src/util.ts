import { exec as _exec } from '@actions/exec';
import { ExecOptions as _ExecOptions } from '@actions/exec/lib/interfaces';
import log from './utils/logger';
import { getInput, startGroup, endGroup } from '@actions/core';
import { join } from 'path';
import fs from 'fs';
import { promisify } from 'util';
import findUp = require('find-up');

const readFileAsync = promisify(fs.readFile);

export type ExecOptions = _ExecOptions;

export type ExecResult = {
  readonly code: number;
  readonly stdout: string;
  readonly stderr: string;
};

export async function exec(
  cmd: string,
  args: string[],
  options?: ExecOptions
): Promise<ExecResult> {
  let stdout = '';
  let stderr = '';
  let code: number;

  try {
    startGroup(`${cmd} ${args.join(' ')}`);
    code = await _exec(cmd, args, {
      ...options,
      listeners: {
        stdout: (data: Buffer) => {
          stdout += data.toString();
        },
        stderr: (data: Buffer) => {
          stderr += data.toString();
        },
      },
    });
  } finally {
    endGroup();
  }
  if (code) {
    log.error({ code });
    throw new Error('Command failed');
  }

  return { code, stdout, stderr };
}

export function isDryRun(): boolean {
  const val = getInput('dry-run');
  return !!val && val === 'true';
}

/**
 * Get environment variable or empty string.
 * Used for easy mocking.
 * @param key variable name
 */
export function getEnv(key: string): string {
  return process.env[key] ?? '';
}

export async function readJson<T = unknown>(file: string): Promise<T> {
  const path = join(getEnv('GITHUB_WORKSPACE'), file);
  const res = await import(path);
  // istanbul ignore next
  return res?.default ?? res;
}

export async function readFile(file: string): Promise<string> {
  const path = join(getEnv('GITHUB_WORKSPACE'), file);
  return await readFileAsync(path, 'utf8');
}

export const MultiArgsSplitRe = /\s*(?:[;,]|$)\s*/;

export function getArg(name: string, opts?: { required?: boolean }): string;
export function getArg(
  name: string,
  opts?: { required?: boolean; multi: true }
): string[];
export function getArg(
  name: string,
  opts?: { required?: boolean; multi?: boolean }
): string | string[];
export function getArg(
  name: string,
  opts?: { required?: boolean; multi?: boolean }
): string | string[] {
  const val = getInput(name, opts);
  return opts?.multi ? val.split(MultiArgsSplitRe).filter(Boolean) : val;
}

let _pkg: Promise<string | undefined>;

/**
 * Resolve path for a file relative to renovate root directory (our package.json)
 * @param file a file to resolve
 */
export async function resolveFile(file: string): Promise<string> {
  if (!_pkg) {
    _pkg = findUp('package.json', { cwd: __dirname, type: 'file' });
  }
  const pkg = await _pkg;
  // istanbul ignore if
  if (!pkg) {
    throw new Error('Missing package.json');
  }
  return join(pkg, '../', file);
}
