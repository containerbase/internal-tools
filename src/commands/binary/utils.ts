import { getInput } from '@actions/core';
import is from '@sindresorhus/is';
import { getDefaultVersioning } from 'renovate/dist/modules/datasource';
import { getArch, getArg, getDistro, isDryRun, readJson } from '../../util';
import { readDockerConfig } from '../../utils/config';
import {
  DockerPlatform,
  dockerBuildx,
  dockerRun,
} from '../../utils/docker/common';
import log from '../../utils/logger';
import type { BinaryBuilderConfig, ConfigFile } from '../../utils/types';

export async function getConfig(): Promise<BinaryBuilderConfig> {
  const configFile = getInput('config') || 'builder.json';

  const cfg = await readJson<ConfigFile>(configFile);

  if (!is.object(cfg)) {
    throw new Error('missing-config');
  }

  if (!is.string(cfg.image)) {
    cfg.image = getInput('image', { required: true });
  }

  if (!is.string(cfg.buildArg)) {
    cfg.buildArg = cfg.image.toUpperCase() + '_VERSION';
  }

  await readDockerConfig(cfg);

  return {
    ...cfg,
    ignoredVersions: cfg.ignoredVersions ?? [],
    dryRun: isDryRun(),
    lastOnly: getArg('last-only') == 'true',
    buildArgs: getArg('build-args', { multi: true }),
    versioning: cfg.versioning ?? getDefaultVersioning(cfg.datasource),
  } as BinaryBuilderConfig;
}

export async function createBuilderImage(
  ws: string,
  { buildArgs }: BinaryBuilderConfig
): Promise<void> {
  log('Creating builder image');
  const args = [
    'build',
    '--load',
    '-t',
    'builder',
    '--build-arg',
    `DISTRO=${getDistro()}`,
  ];
  const arch = getArch();
  if (is.nonEmptyString(arch)) {
    args.push('--platform', DockerPlatform[arch]);
  }
  if (is.nonEmptyArray(buildArgs)) {
    args.push(...buildArgs.map((b) => `--build-arg=${b}`));
  }
  await dockerBuildx(...args, ws);
}

export async function runBuilder(ws: string, version: string): Promise<void> {
  const args = ['--name', 'builder', '--volume', `${ws}/.cache:/cache`];
  const arch = getArch();
  if (is.nonEmptyString(arch)) {
    args.push('--platform', DockerPlatform[arch]);
  }
  await dockerRun(...args, 'builder', version);
}
