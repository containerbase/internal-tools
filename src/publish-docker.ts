import { getInput } from '@actions/core';
import { exec } from './util';
import chalk from 'chalk';
import log from 'fancy-log';
import { getRemoteImageId, getLocalImageId } from './utils/docker';

export const MultiArgsSplitRe = /\s*(?:;|$)\s*/;
export async function run(): Promise<void> {
  const dryRun = !!getInput('dry-run');
  const image = getInput('image');
  let tags = getInput('tags')
    ?.split(MultiArgsSplitRe)
    .filter(Boolean);

  if (!tags?.length) {
    tags = ['latest'];
  }

  if (!image) {
    throw new Error('Missing image');
  }

  for (const tag of tags) {
    const fullName = `${image}:${tag}`;
    log.info(chalk.blue('Processing image:'), chalk.yellow(fullName));

    log('Fetch new id');
    const newId = await getLocalImageId(image, tag);

    log('Fetch old id');
    const oldId = await getRemoteImageId(image, tag);

    if (oldId === newId) {
      log('Image uptodate, no push nessessary: ', chalk.yellow(oldId));
      continue;
    }

    log('Publish new image', `${oldId} > ${newId}`);
    if (dryRun) {
      log.warn(chalk.yellow('DRY-RUN: Would push: '), fullName);
    } else {
      await exec('docker', ['push', fullName]);
    }
    log.info(chalk.blue('Processing image finished: ', newId));
  }
}
