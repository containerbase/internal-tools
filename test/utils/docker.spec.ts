import * as nock from 'nock';
import * as _utils from '../../src/util';
import {
  DockerContentType,
  build,
  getAuthHeaders,
  getLocalImageId,
  getRemoteImageId,
  publish,
  splitImage,
} from '../../src/utils/docker';
import { ExecError } from '../../src/utils/types';
import { mocked } from '../utils';

jest.mock('delay', () => () => Promise.resolve());
jest.mock('../../src/util');

const utils = mocked(_utils);
const res = { code: 0, stdout: '', stderr: '' };

const imagePrefix = 'renovate';
const digest =
  'sha256:d694b03ba0df63ac9b27445e76657d4ed62898d721b997372aab150ee84e07a1';
const tag = 'latest';
const realm = 'https://auth.docker.io';
const registry = 'https://index.docker.io';

describe('utils/docker', () => {
  beforeEach(() => {
    jest.resetAllMocks();
    nock.cleanAll();
  });

  afterEach(() => {
    nock.abortPendingRequests();
  });

  describe('splitImage', () => {
    it('works', () => {
      expect(splitImage('renovatebot/base')).toEqual({
        registry: 'index.docker.io',
        repository: 'renovatebot/base',
      });
      expect(splitImage('ghcr.io/containerbase/base')).toEqual({
        registry: 'ghcr.io',
        repository: 'containerbase/base',
      });
    });
  });

  describe('getAuthHeaders', () => {
    const image = 'renovate/base';
    const headers = {
      'www-authenticate': `Bearer realm="${realm}/token",service="registry.docker.io",scope="repository:${image}:pull"`,
    };

    it('works', async () => {
      nock(registry).get('/v2/').reply(200, {}, headers);
      nock(realm)
        .get(`/token?service=registry.docker.io&scope=repository:${image}:pull`)
        .reply(200, { access_token: 'XXX' });
      expect(await getAuthHeaders(registry.slice(8), image)).toEqual({
        authorization: 'Bearer XXX',
      });

      expect(nock.isDone()).toBe(true);
    });

    it('fails with auth', async () => {
      nock(registry).get('/v2/').reply(200, {}, headers);
      nock(realm)
        .get(`/token?service=registry.docker.io&scope=repository:${image}:pull`)
        .reply(200, {});
      await expect(getAuthHeaders(registry.slice(8), image)).rejects.toThrow(
        'Failed to obtain docker registry token'
      );

      expect(nock.isDone()).toBe(true);
    });

    it('fails with http error', async () => {
      nock(registry).get('/v2/').reply(200, {}, headers);
      nock(realm)
        .get(`/token?service=registry.docker.io&scope=repository:${image}:pull`)
        .reply(418);
      await expect(getAuthHeaders(registry.slice(8), image)).rejects.toThrow(
        'Failed to obtain docker registry token'
      );

      expect(nock.isDone()).toBe(true);
    });
  });

  describe('getRemoteImageId', () => {
    const image = 'renovate/base';

    it('works', async () => {
      nock(registry)
        .get('/v2/')
        .reply(200, {})
        .get(`/v2/${image}/manifests/${tag}`)
        .reply(
          200,
          {
            config: {
              digest,
            },
          },
          { 'content-type': DockerContentType.ManifestV2 }
        );
      expect(await getRemoteImageId(registry.slice(8), image)).toEqual(digest);
      expect(nock.isDone()).toBe(true);
    });

    it('return <none> on 404', async () => {
      nock(registry)
        .get('/v2/')
        .reply(200)
        .get(`/v2/${image}/manifests/${tag}`)
        .reply(404);

      expect(await getRemoteImageId(registry.slice(8), image)).toBe('<none>');
      expect(nock.isDone()).toBe(true);
    });

    it('return <error> on manifest v1', async () => {
      nock(registry)
        .get('/v2/')
        .reply(200, {})
        .get(`/v2/${image}/manifests/${tag}`)
        .reply(200, {}, { 'content-type': DockerContentType.ManifestV1 });

      expect(await getRemoteImageId(registry.slice(8), image)).toBe('<error>');
      expect(nock.isDone()).toBe(true);
    });

    it('throws', async () => {
      nock(registry)
        .get('/v2/')
        .reply(200)
        .get(`/v2/${image}/manifests/${tag}`)
        .reply(418);
      await expect(getRemoteImageId(registry.slice(8), image)).rejects.toThrow(
        'Could not find remote image id'
      );
      expect(nock.isDone()).toBe(true);
    });

    it('throws unsupported', async () => {
      nock(registry)
        .get('/v2/')
        .reply(200, {})
        .get(`/v2/${image}/manifests/${tag}`)
        .reply(200, {}, { 'content-type': 'unsupported' });

      await expect(getRemoteImageId(registry.slice(8), image)).rejects.toThrow(
        'Could not find remote image id'
      );
      expect(nock.isDone()).toBe(true);
    });
  });

  describe('getLocalImageId', () => {
    const image = 'renovate/base';

    it('works', async () => {
      utils.exec.mockResolvedValueOnce({
        ...res,
        stdout: `'${digest}'\n"`,
      });

      expect(await getLocalImageId(image)).toEqual(digest);
    });

    it('throws', async () => {
      utils.exec.mockResolvedValueOnce({
        ...res,
        stdout: '',
      });

      await expect(getLocalImageId(image)).rejects.toThrow(
        'Could not find local image id'
      );
    });
  });

  describe('build', () => {
    const image = 'base';
    const cache = 'docker-build-cache';

    it('works', async () => {
      utils.exec.mockResolvedValueOnce({
        ...res,
      });

      await build({ imagePrefix, image, buildArgs: [] });
      expect(utils.exec.mock.calls).toMatchSnapshot();
    });

    it('uses cache', async () => {
      utils.exec.mockResolvedValueOnce({
        ...res,
      });

      await build({ imagePrefix, image, cache });
      expect(utils.exec.mock.calls).toMatchSnapshot();
    });

    it('uses cache (dry-run)', async () => {
      utils.exec.mockResolvedValueOnce({
        ...res,
      });

      await build({
        imagePrefix,
        image,
        cache,
        cacheTags: ['dummy'],
        dryRun: true,
        buildArgs: ['IMAGE=slim'],
      });
      expect(utils.exec.mock.calls).toMatchSnapshot();
    });

    it('retries', async () => {
      utils.exec.mockRejectedValueOnce(
        new ExecError(1, 'failed', 'unexpected status: 400 Bad Request', '')
      );
      utils.exec.mockResolvedValueOnce({
        ...res,
      });

      await build({ imagePrefix, image });
      expect(utils.exec.mock.calls).toMatchSnapshot();
    });

    it('throws', async () => {
      utils.exec.mockRejectedValueOnce(
        new ExecError(1, 'failed', 'unexpected status: 400 Bad Request', '')
      );
      utils.exec.mockRejectedValueOnce(new Error('failure'));

      await expect(build({ imagePrefix, image })).rejects.toThrow('failure');

      expect(utils.exec.mock.calls).toMatchSnapshot();
    });
  });

  describe('publish', () => {
    const image = 'base';

    beforeEach(() => {
      nock(registry)
        .get('/v2/')
        .reply(200, {})
        .get(`/v2/renovate/${image}/manifests/${tag}`)
        .reply(
          200,
          {
            config: {
              digest,
            },
          },
          { 'content-type': DockerContentType.ManifestV2 }
        );
    });

    it('works', async () => {
      utils.exec.mockResolvedValueOnce({
        ...res,
        stdout:
          'sha256:d694b03ba0df63ac9b27445e76657d4ed62898d721b997372aab150ee84e07a2',
      });
      utils.exists.mockResolvedValueOnce(true);

      await publish({ imagePrefix, image, tag });
      expect(utils.exec.mock.calls).toMatchSnapshot();

      expect(nock.isDone()).toBe(true);
    });

    it('works (dry-run)', async () => {
      utils.exec.mockResolvedValueOnce({
        ...res,
        stdout:
          'sha256:d694b03ba0df63ac9b27445e76657d4ed62898d721b997372aab150ee84e07a2',
      });

      await publish({ imagePrefix, image, tag, dryRun: true });
      expect(utils.exec.mock.calls).toMatchSnapshot();

      expect(nock.isDone()).toBe(true);
    });

    it('uptodate', async () => {
      utils.exec.mockResolvedValueOnce({
        ...res,
        stdout: digest,
      });

      await publish({ imagePrefix, image, tag });
      expect(utils.exec.mock.calls).toMatchSnapshot();

      expect(nock.isDone()).toBe(true);
    });

    it('skips signing', async () => {
      utils.exec.mockResolvedValueOnce({
        ...res,
        stdout:
          'sha256:d694b03ba0df63ac9b27445e76657d4ed62898d721b997372aab150ee84e07a2',
      });
      utils.exists.mockResolvedValue(false);

      await publish({ imagePrefix, image, tag });
      expect(utils.exec.mock.calls).toMatchSnapshot();

      expect(nock.isDone()).toBe(true);
    });
  });
});
