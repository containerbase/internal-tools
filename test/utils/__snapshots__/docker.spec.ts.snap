// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` build does platform things) 1`] = `
Array [
  Array [
    "docker",
    Array [
      "buildx",
      "build",
      "--tag=renovate/base:latest",
      "--load",
      "--platform=linux/arm64",
      "--cache-from=renovate/docker-build-cache:base-latest",
      "--cache-from=renovate/docker-build-cache:base-dummy",
      ".",
    ],
  ],
]
`;

exports[` build multiplatform (dry-run) 1`] = `
Array [
  Array [
    "docker",
    Array [
      "buildx",
      "build",
      "--tag=renovate/base:latest",
      "--platform=linux/amd64",
      "--platform=linux/arm64",
      "--cache-from=renovate/docker-build-cache:base-latest",
      ".",
    ],
  ],
]
`;

exports[` build multiplatform 1`] = `
Array [
  Array [
    "docker",
    Array [
      "buildx",
      "build",
      "--tag=renovate/base:latest",
      "--push",
      "--platform=linux/amd64",
      "--platform=linux/arm64",
      "--cache-from=renovate/docker-build-cache:base-latest",
      "--cache-to=type=registry,ref=renovate/docker-build-cache:base-latest,mode=max",
      ".",
    ],
  ],
]
`;

exports[` build retries 1`] = `
Array [
  Array [
    "docker",
    Array [
      "buildx",
      "build",
      "--tag=renovate/base:latest",
      "--load",
      ".",
    ],
  ],
  Array [
    "docker",
    Array [
      "buildx",
      "build",
      "--tag=renovate/base:latest",
      "--load",
      ".",
    ],
  ],
]
`;

exports[` build throws 1`] = `
Array [
  Array [
    "docker",
    Array [
      "buildx",
      "build",
      "--tag=renovate/base:latest",
      "--load",
      ".",
    ],
  ],
  Array [
    "docker",
    Array [
      "buildx",
      "build",
      "--tag=renovate/base:latest",
      "--load",
      ".",
    ],
  ],
]
`;

exports[` build uses cache (dry-run) 1`] = `
Array [
  Array [
    "docker",
    Array [
      "buildx",
      "build",
      "--tag=renovate/base:latest",
      "--load",
      "--build-arg=IMAGE=slim",
      "--cache-from=renovate/docker-build-cache:base-latest",
      "--cache-from=renovate/docker-build-cache:base-dummy",
      ".",
    ],
  ],
]
`;

exports[` build uses cache 1`] = `
Array [
  Array [
    "docker",
    Array [
      "buildx",
      "build",
      "--tag=renovate/base:latest",
      "--load",
      "--cache-from=renovate/docker-build-cache:base-latest",
      "--cache-to=type=registry,ref=renovate/docker-build-cache:base-latest,mode=max",
      ".",
    ],
  ],
]
`;

exports[` build works 1`] = `
Array [
  Array [
    "docker",
    Array [
      "buildx",
      "build",
      "--tag=renovate/base:latest",
      "--load",
      ".",
    ],
  ],
]
`;

exports[` publish uptodate 1`] = `
Array [
  Array [
    "docker",
    Array [
      "inspect",
      "--format='{{.Id}}'",
      "renovate/base:latest",
    ],
  ],
]
`;

exports[` publish works (dry-run) 1`] = `
Array [
  Array [
    "docker",
    Array [
      "inspect",
      "--format='{{.Id}}'",
      "renovate/base:latest",
    ],
  ],
]
`;

exports[` publish works 1`] = `
Array [
  Array [
    "docker",
    Array [
      "inspect",
      "--format='{{.Id}}'",
      "renovate/base:latest",
    ],
  ],
  Array [
    "docker",
    Array [
      "push",
      "renovate/base:latest",
    ],
  ],
]
`;
